modules/table/src/main/java/org/apache/ignite/internal/table/distributed/gc/MvGc.java

```diff
@@ -0,0 +1,249 @@
+     *
+     * <p>Should be called before rebalancing/closing/destroying the storage.
+     *
+     * @param tablePartitionId Table partition ID.
+     * @return Storage garbage collection completion future.
+     * @throws IgniteInternalException with {@link Gc#CLOSED_ERR} If the garbage collector is closed.
+     */
+    public CompletableFuture<Void> removeStorage(TablePartitionId tablePartitionId) {
+        return inBusyLock(() -> {
+            GcStorageHandler removed = storageHandlerByPartitionId.remove(tablePartitionId);
+
+            if (removed == null) {
+                return completedFuture(null);
+            }
+
+            CompletableFuture<Void> gcInProgressFuture = removed.gcInProgressFuture.get();
+
+            return gcInProgressFuture == null ? completedFuture(null) : gcInProgressFuture;
+        });
+    }
+
+    /**
+     * Updates the new watermark only if it is larger than the current low watermark.
+     *
+     * <p>If the update is successful, it will schedule a new garbage collection for all storages.
+     *
+     * @param newLwm New low watermark.
+     * @throws IgniteInternalException with {@link Gc#CLOSED_ERR} If the garbage collector is closed.
+     */
+    public void updateLowWatermark(HybridTimestamp newLwm) {
+        inBusyLock(() -> {
+            HybridTimestamp updatedLwm = lowWatermarkReference.updateAndGet(currentLwm -> {
+                if (currentLwm == null) {
+                    return newLwm;
+                }
+
+                // Update only if the new one is greater than the current one.
+                return newLwm.compareTo(currentLwm) > 0 ? newLwm : currentLwm;
+            });
+
+            // If the new watermark is smaller than the current one or has been updated in parallel, then we do nothing.
+            if (updatedLwm != newLwm) {
+                return;
+            }
+
+            executor.submit(() -> inBusyLock(this::initNewGcBusy));
+        });
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (!closeGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+
+        shutdownAndAwaitTermination(executor, 10, TimeUnit.SECONDS);
+    }
+
+    private void initNewGcBusy() {
+        storageHandlerByPartitionId.keySet().forEach(this::scheduleGcForStorage);
+    }
+
+    private void scheduleGcForStorage(TablePartitionId tablePartitionId) {
+        executor.submit(() -> inBusyLock(() -> {
+            GcStorageHandler storageHandler = storageHandlerByPartitionId.compute(tablePartitionId, (id, gcStorageHandler) -> {
+                if (gcStorageHandler == null) {
+                    // Storage has been removed from garbage collection.
+                    return gcStorageHandler;
+                }
+
+                boolean casResult = gcStorageHandler.gcInProgressFuture.compareAndSet(null, new CompletableFuture<>());
```

Comment:
You could have instantiated the future in advance, this way there would be no need to read it from the map later and assert that it's not null.

Reply:
I don't quite understand you at the moment.

Reply:
Can you please clarify, what exactly you don't understand?

Reply:
We discussed personally, perhaps fixed.